File: pbark/pbark/interpreter/dog.py
from __future__ import annotations

from pbark.config.config_loader import ConfigLoader
from pbark.config.dog_topics import DogTopics
from pbark.errors import BarkError
from pbark.interpreter import bark_value as bv


class Dog:
    def __init__(self) -> None:
        self.food = 0.0
        self.items = 0.0
        self.inventory: list[str] = []
        self.age = 0.0
        self.name = ""
        self.description = ""
        self.color = ""
        self.happy = False
        self.fed = False

    @staticmethod
    def fresh() -> Dog:
        return Dog()

    def get(self, topic: str, line: int) -> bv.BarkValue:
        field = DogTopics.resolve(topic)
        if field is None:
            raise BarkError(
                line,
                f'This dog doesn\'t track "{topic}". Try one of: {DogTopics.CANONICAL}.',
            )
        if field == DogTopics.FOOD:
            return bv.of(self.food)
        if field == DogTopics.ITEMS:
            return bv.of(self.items)
        if field == DogTopics.INVENTORY:
            return bv.of(self._format_inventory())
        if field == DogTopics.AGE:
            return bv.of(self.age)
        if field == DogTopics.NAME:
            return bv.of(self.name)
        if field == DogTopics.DESCRIPTION:
            return bv.of(self.description)
        if field == DogTopics.COLOR:
            return bv.of(self.color)
        if field == DogTopics.HAPPY:
            return bv.of(self.happy)
        if field == DogTopics.FED:
            return bv.of(self.fed)
        raise BarkError(line, f'Unknown dog field "{field}".')

    def set(self, topic: str, value: bv.BarkValue, line: int) -> None:
        field = DogTopics.resolve(topic)
        if field is None:
            raise BarkError(
                line,
                f'This dog doesn\'t track "{topic}". Try one of: {DogTopics.CANONICAL}.',
            )
        if field == DogTopics.FOOD:
            self.food = self._as_number(value, line)
        elif field == DogTopics.ITEMS:
            self.items = self._as_number(value, line)
        elif field == DogTopics.INVENTORY:
            self._set_inventory(value, line)
        elif field == DogTopics.AGE:
            self.age = self._as_number(value, line)
        elif field == DogTopics.NAME:
            self.name = self._as_string(value, line)
        elif field == DogTopics.DESCRIPTION:
            self.description = self._as_string(value, line)
        elif field == DogTopics.COLOR:
            self.color = self._as_string(value, line)
        elif field == DogTopics.HAPPY:
            self.happy = self._as_bool(value, line)
        elif field == DogTopics.FED:
            self.fed = self._as_bool(value, line)
        else:
            raise BarkError(line, f'Unknown dog field "{field}".')

    def count_object(self, object_name: str) -> int:
        key = ConfigLoader.normalise(object_name)
        return sum(1 for held in self.inventory if held == key)

    def add_held_object(self, object_name: str) -> None:
        self.inventory.append(ConfigLoader.normalise(object_name))

    def remove_held_object(self, object_name: str) -> None:
        key = ConfigLoader.normalise(object_name)
        for i, held in enumerate(self.inventory):
            if held == key:
                del self.inventory[i]
                return

    def remove_any_held_object(self) -> None:
        if self.inventory:
            del self.inventory[-1]

    def _format_inventory(self) -> str:
        if not self.inventory:
            return ""
        return ", ".join(h.replace("_", " ") for h in self.inventory)

    def _set_inventory(self, value: bv.BarkValue, line: int) -> None:
        self.inventory.clear()
        if isinstance(value, bv.BarkNull):
            return
        if not isinstance(value, bv.BarkString):
            raise BarkError(
                line,
                f"She can't carry {bv.describe(value)} in her mouth. "
                "List things like ball, stick (comma-separated words).",
            )
        if not value.value.strip():
            return
        for part in value.value.split(","):
            trimmed = part.strip()
            if trimmed:
                self.inventory.append(ConfigLoader.normalise(trimmed.replace(" ", "_")))

    @staticmethod
    def _as_number(value: bv.BarkValue, line: int) -> float:
        if isinstance(value, bv.BarkNumber):
            return value.value
        raise BarkError(line, f"That needs a number, not {bv.describe(value)}.")

    @staticmethod
    def _as_string(value: bv.BarkValue, line: int) -> str:
        if isinstance(value, bv.BarkString):
            return value.value
        raise BarkError(line, f"That needs words on her tag, not {bv.describe(value)}.")

    @staticmethod
    def _as_bool(value: bv.BarkValue, line: int) -> bool:
        if isinstance(value, bv.BarkBoolean):
            return value.value
        raise BarkError(line, f"Wag yes or naw no. Not {bv.describe(value)}.")

File: src/main/java/dev/klomptech/jbark/interpreter/Dog.java
package dev.klomptech.jbark.interpreter;

import dev.klomptech.jbark.config.ConfigLoader;
import dev.klomptech.jbark.config.DogTopics;
import dev.klomptech.jbark.errors.BarkError;
import java.util.ArrayList;
import java.util.List;

// One breed's runtime state: food, toys, name, and the rest
public final class Dog {

  private double food;
  private double items;
  private final List<String> inventory = new ArrayList<>();
  private double age;
  private String name = "";
  private String description = "";
  private String color = "";
  private boolean happy;
  private boolean fed;

  public static Dog fresh() {
    return new Dog();
  }

  public BarkValue get(final String topic, final int line) throws BarkError {
    String field = DogTopics.resolve(topic);
    if (field == null) {
      throw new BarkError(
          line,
          "This dog doesn't track \"" + topic + "\". Try one of: " + DogTopics.CANONICAL + ".");
    }
    return switch (field) {
      case DogTopics.FOOD -> BarkValue.of(food);
      case DogTopics.ITEMS -> BarkValue.of(items);
      case DogTopics.INVENTORY -> BarkValue.of(formatInventory());
      case DogTopics.AGE -> BarkValue.of(age);
      case DogTopics.NAME -> BarkValue.of(name);
      case DogTopics.DESCRIPTION -> BarkValue.of(description);
      case DogTopics.COLOR -> BarkValue.of(color);
      case DogTopics.HAPPY -> BarkValue.of(happy);
      case DogTopics.FED -> BarkValue.of(fed);
      default -> throw new BarkError(line, "Unknown dog field \"" + field + "\".");
    };
  }

  public void set(final String topic, final BarkValue value, final int line) throws BarkError {
    String field = DogTopics.resolve(topic);
    if (field == null) {
      throw new BarkError(
          line,
          "This dog doesn't track \"" + topic + "\". Try one of: " + DogTopics.CANONICAL + ".");
    }
    switch (field) {
      case DogTopics.FOOD -> food = asNumber(value, line);
      case DogTopics.ITEMS -> items = asNumber(value, line);
      case DogTopics.INVENTORY -> setInventory(value, line);
      case DogTopics.AGE -> age = asNumber(value, line);
      case DogTopics.NAME -> name = asString(value, line);
      case DogTopics.DESCRIPTION -> description = asString(value, line);
      case DogTopics.COLOR -> color = asString(value, line);
      case DogTopics.HAPPY -> happy = asBool(value, line);
      case DogTopics.FED -> fed = asBool(value, line);
      default -> throw new BarkError(line, "Unknown dog field \"" + field + "\".");
    }
  }

  public int countObject(final String object) {
    String key = ConfigLoader.normalise(object);
    int count = 0;
    for (String held : inventory) {
      if (held.equals(key)) {
        count++;
      }
    }
    return count;
  }

  public void addHeldObject(final String object) {
    inventory.add(ConfigLoader.normalise(object));
  }

  public void removeHeldObject(final String object) {
    String key = ConfigLoader.normalise(object);
    for (int i = 0; i < inventory.size(); i++) {
      if (inventory.get(i).equals(key)) {
        inventory.remove(i);
        return;
      }
    }
  }

  public void removeAnyHeldObject() {
    if (!inventory.isEmpty()) {
      inventory.remove(inventory.size() - 1);
    }
  }

  private String formatInventory() {
    if (inventory.isEmpty()) {
      return "";
    }
    List<String> parts = new ArrayList<>();
    for (String held : inventory) {
      parts.add(held.replace('_', ' '));
    }
    return String.join(", ", parts);
  }

  private void setInventory(final BarkValue value, final int line) throws BarkError {
    inventory.clear();
    if (value instanceof BarkValue.BarkNull) {
      return;
    }
    if (!(value instanceof BarkValue.BarkString text)) {
      throw new BarkError(
          line,
          "She can't carry "
              + BarkValue.describe(value)
              + " in her mouth. "
              + "List things like ball, stick (comma-separated words).");
    }
    if (text.value().isBlank()) {
      return;
    }
    for (String part : text.value().split(",")) {
      String trimmed = part.trim();
      if (!trimmed.isEmpty()) {
        inventory.add(ConfigLoader.normalise(trimmed.replace(' ', '_')));
      }
    }
  }

  private static double asNumber(final BarkValue value, final int line) throws BarkError {
    if (value instanceof BarkValue.BarkNumber n) {
      return n.value();
    }
    throw new BarkError(line, "That needs a number, not " + BarkValue.describe(value) + ".");
  }

  private static String asString(final BarkValue value, final int line) throws BarkError {
    if (value instanceof BarkValue.BarkString s) {
      return s.value();
    }
    throw new BarkError(line, "That needs words on her tag, not " + BarkValue.describe(value) + ".");
  }

  private static boolean asBool(final BarkValue value, final int line) throws BarkError {
    if (value instanceof BarkValue.BarkBoolean b) {
      return b.value();
    }
    throw new BarkError(line, "Wag yes or naw no. Not " + BarkValue.describe(value) + ".");
  }
}
